### ISSUE TITLE
Macros defined in `global_asm!` blocks aren't global

### ISSUE THREAD

**Issue by gwilymk** [rust-lang/rust#145409]
### Code

I tried this code:

```rust
use std::arch::global_asm;

global_asm!(
    "
    .macro foo
    .endm
    "
);

global_asm!(
    "
    foo
    "
);

```

I expected to see this happen: No compiler error

Instead, this happened: Compiler error

```
   Compiling playground v0.0.1 (/playground)
error: invalid instruction mnemonic 'foo'
  |
note: instantiated into assembly here
 --> <inline asm>:3:5
  |
3 |     foo
  |     ^^^

error: could not compile `playground` (lib) due to 1 previous error
```

Playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=700d72725134c95f3b12b26587e90af4

### Version it worked on

It most recently worked on: Rust 1.89.0 and Rust 1.90.0 beta

Also works on nightly-2025-08-12

### Version with regression

`rustc --version --verbose`:
```
1.91.0-nightly (2025-08-13 3672a55b7cfd0a12e709)
```

@rustbot modify labels: +regression-from-stable-to-nightly -regression-untriaged

**Comment by hanna-kruppe** [comment]
I don’t think there’s any guarantees about the order in which different inline assembly blocks are processed by the assembler. The reference doesn’t explicit mention assembler macros specifically but in general [stateful directives must be reverted before the end of a block](https://doc.rust-lang.org/nightly/reference/inline-assembly.html#r-asm.directives.stateful) and I guess this covers macros as well.

**Comment by Noratrieb** [comment]
Yes, inline assembly must not share any assembler state as the order of everything is undefined.
You can use a Rust macro to share code between different inline assembly invocations (which is a bit less ergonomic but works reliably).
